lib/models/per-bundle-addon-cache/index.js

```diff
@@ -0,0 +1,387 @@
+          this.numProxies++;
+          return getAddonProxy(addonCacheEntry, parent);
+        } else {
+          // the cache entry was created 'empty' by an earlier call, indicating
+          // an addon that is used in a lazy engine but also used in the project,
+          // and we're now creating the instance for the project.
+          // Fill in the entry and return the new instance.
+          logger.debug(`About to fill in REGULAR ADDON EXISTING cache entry for addon at: ${addonPkgInfo.realPath}`);
+          this.numAddonInstances++;
+          addonInstance = addonPkgInfo.constructAddonInstance(parent, this.project);
+          addonCacheEntry.targetInstance = addonInstance; // cache BEFORE initializing child addons
+          addonPkgInfo.initChildAddons(addonInstance);
+          return addonInstance;
+        }
+      }
+
+      // There is no entry for this addon in the bundleRoot's addon cache. Create a new
+      // instance, cache it in the addon cache, and return it.
+      logger.debug(`About to construct REGULAR ADDON NEW cache entry for addon at: ${addonPkgInfo.realPath}`);
+      this.numAddonInstances++;
+      addonInstance = addonPkgInfo.constructAddonInstance(parent, this.project);
+      addonCacheEntry = this.createAddonCacheEntry(addonInstance, addonPkgInfo.realPath);
+      addonInstanceCache.set(addonPkgInfo.realPath, addonCacheEntry); // cache BEFORE initializing child addons
+      addonPkgInfo.initChildAddons(addonInstance);
+      return addonInstance;
+    } else {
+      // The bundleRoot is not the project but the project bundles the addon too, so
+      // the cache entry needs to go in the project's cache. Get/create an empty
+      // cache entry and return a proxy to it. The project will set the instance later
+      // (see above).
+      let bundleRootCacheEntry = this.bundleRootCache.get(PROJECT_BUNDLE_ROOT_NAME);
+      let addonCacheEntry = bundleRootCacheEntry.addonInstanceCache.get(addonPkgInfo.realPath);
+      if (!addonCacheEntry) {
+        logger.debug(`About to construct REGULAR ADDON EMPTY cache entry for addon at: ${addonPkgInfo.realPath}`);
+        addonCacheEntry = this.createAddonCacheEntry(null, addonPkgInfo.realPath);
+        bundleRootCacheEntry.addonInstanceCache.set(addonPkgInfo.realPath, addonCacheEntry);
+      }
+
+      logger.debug(`About to construct REGULAR ADDON PROXY for EMPTY addon at: ${addonPkgInfo.realPath}`);
+      this.numProxies++;
+      return getAddonProxy(addonCacheEntry, parent);
+    }
+  }
+
+  /**
+   * Called by engine-addon when it needs to retrieve the real child addons for a given
+   * object (i.e., during ancestorHostAddons), even though the object might be a proxy.
+   * 
+   * @method getAddonAddons
+   * @param {Addon|Proxy} addonInstance an instance of an addon or a Proxy generated by this class
+   * @return {Array} an array of addon instances
+   */
+  getAddonAddons(addonInstance) {
```

Comment:
This is the method that `ember-engines` should call inside `ancestorHostAddons`, instead of futzing around with knowing that `targetInstance` exists or replacing the method with a custom one. I'll add a comment.